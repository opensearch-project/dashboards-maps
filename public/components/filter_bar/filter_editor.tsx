/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiButton,
  EuiButtonEmpty,
  EuiCodeEditor,
  EuiFieldText,
  EuiFlexGroup,
  EuiFlexItem,
  EuiForm,
  EuiFormRow,
  EuiPopoverTitle,
  EuiSpacer,
} from '@elastic/eui';
import React, { useState } from 'react';
import { i18n } from '@osd/i18n';

interface Props {
  content: string;
  label: string;
  onSubmit: (content: string, label: string) => void;
  onCancel: () => void;
}

const isFilterValid = (content: string) => {
  try {
    return Boolean(JSON.parse(content));
  } catch (e) {
    return false;
  }
};

export const FilterEditor = ({ content, label, onSubmit, onCancel }: Props) => {
  const [filterLabel, setFilterLabel] = useState<string>(label);
  const [filterContent, setFilterContent] = useState<string>(content);

  const renderEditor = () => {
    return (
      <EuiFormRow
        label={i18n.translate('maps.filter.filterEditor.parameters', {
          defaultMessage: 'Spatial filter parameters',
        })}
        fullWidth={true}
      >
        <EuiCodeEditor
          value={filterContent}
          onChange={setFilterContent}
          mode="json"
          width="100%"
          height="250px"
        />
      </EuiFormRow>
    );
  };
  return (
    <div>
      <EuiPopoverTitle>
        <EuiFlexGroup alignItems="baseline" responsive={false}>
          <EuiFlexItem>{'Edit Filter'}</EuiFlexItem>
        </EuiFlexGroup>
      </EuiPopoverTitle>

      <div className="globalFilterItem__editorForm">
        <EuiForm>
          {renderEditor()}
          <div>
            <EuiSpacer size="m" />
            <EuiFormRow
              fullWidth={true}
              label={i18n.translate('maps.filter.filterEditor.createCustomLabelInputLabel', {
                defaultMessage: 'Custom label',
              })}
            >
              <EuiFieldText
                fullWidth={true}
                value={filterLabel}
                onChange={(event) => setFilterLabel(event.target.value)}
              />
            </EuiFormRow>
          </div>
          <EuiSpacer size="m" />
          <EuiFlexGroup direction="rowReverse" alignItems="center" responsive={false}>
            <EuiFlexItem grow={false}>
              <EuiButton
                fill
                onClick={() => onSubmit(filterContent, filterLabel)}
                isDisabled={!isFilterValid(filterContent)}
                data-test-subj="saveFilter"
              >
                {'Save'}
              </EuiButton>
            </EuiFlexItem>
            <EuiFlexItem grow={false}>
              <EuiButtonEmpty flush="right" onClick={onCancel} data-test-subj="cancelSaveFilter">
                {'Cancel'}
              </EuiButtonEmpty>
            </EuiFlexItem>
            <EuiFlexItem />
          </EuiFlexGroup>
        </EuiForm>
      </div>
    </div>
  );
};
