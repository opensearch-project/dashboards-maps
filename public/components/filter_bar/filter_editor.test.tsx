/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { shallow } from 'enzyme';
import React from 'react';
import { FilterEditor } from './filter_editor';

it('renders one filter inside filter bar', () => {
  const mockOnSubmitCallback = jest.fn();
  const mockOnCancelCallback = jest.fn();
  const filterEditor = shallow(
    <FilterEditor
      content={JSON.stringify('{}')}
      label={'mylabel'}
      onCancel={mockOnCancelCallback}
      onSubmit={mockOnSubmitCallback}
    />
  );
  expect(filterEditor).toMatchSnapshot();
});
