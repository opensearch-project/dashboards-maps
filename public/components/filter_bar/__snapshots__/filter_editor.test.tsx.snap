// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders one filter inside filter bar 1`] = `
<div>
  <EuiPopoverTitle>
    <EuiFlexGroup
      alignItems="baseline"
      responsive={false}
    >
      <EuiFlexItem>
        Edit Filter
      </EuiFlexItem>
    </EuiFlexGroup>
  </EuiPopoverTitle>
  <div
    className="globalFilterItem__editorForm"
  >
    <EuiForm>
      <EuiFormRow
        describedByIds={Array []}
        display="row"
        fullWidth={true}
        hasChildLabel={true}
        hasEmptyLabelSpace={false}
        label="Spatial filter parameters"
        labelType="label"
      >
        <EuiCodeEditor
          height="250px"
          mode="json"
          onChange={[Function]}
          setOptions={Object {}}
          value="\\"{}\\""
          width="100%"
        />
      </EuiFormRow>
      <div>
        <EuiSpacer
          size="m"
        />
        <EuiFormRow
          describedByIds={Array []}
          display="row"
          fullWidth={true}
          hasChildLabel={true}
          hasEmptyLabelSpace={false}
          label="Custom label"
          labelType="label"
        >
          <EuiFieldText
            fullWidth={true}
            onChange={[Function]}
            value="mylabel"
          />
        </EuiFormRow>
      </div>
      <EuiSpacer
        size="m"
      />
      <EuiFlexGroup
        alignItems="center"
        direction="rowReverse"
        responsive={false}
      >
        <EuiFlexItem
          grow={false}
        >
          <EuiButton
            data-test-subj="saveFilter"
            fill={true}
            isDisabled={false}
            onClick={[Function]}
          >
            Save
          </EuiButton>
        </EuiFlexItem>
        <EuiFlexItem
          grow={false}
        >
          <EuiButtonEmpty
            data-test-subj="cancelSaveFilter"
            flush="right"
            onClick={[MockFunction]}
          >
            Cancel
          </EuiButtonEmpty>
        </EuiFlexItem>
        <EuiFlexItem />
      </EuiFlexGroup>
    </EuiForm>
  </div>
</div>
`;
